### PR TITLE
Introduce TestNG to JUnit migration module [DRAFT]

### PR DESCRIPTION
**Pull request by oxkitsune**
DRAFT

### CODE REVIEW

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+   * migration mode.
+   */
+  public TestNGJUnitMigration() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a new {@link TestNGJUnitMigration} with the specified {@link ErrorProneFlags}.
+   *
+   * @param flags the error-prone flags used to set the migration mode
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java:63]:
```suggestion
   * @param flags the error-prone flags used to set the migration mode.
```

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                });
+
+        metaData
+            .getAnnotation(tree)
+            .filter(annotation -> canMigrateTest(tree, metaData, annotation, state))
+            .ifPresent(
+                annotation -> {
+                  SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+
+                  // migrate arguments
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java:115]: Can omit this? Or make a proper comment. `/* Migrate arguments. */`

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java [hunk 3]
```diff
@@ -704,9 +704,7 @@ Flux<T> after(Mono<T> mono) {
   static final class MonoSingleOptional<T> {
     @BeforeTemplate
     Mono<Optional<T>> before(Mono<T> mono) {
-      return Refaster.anyOf(
-          mono.flux().collect(toOptional()),
-          mono.map(Optional::of).defaultIfEmpty(Optional.empty()));
+      return Refaster.anyOf(mono.flux().collect(toOptional()), mono.flux().collect(toOptional()));
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:707]: This should be reverted.

File: testng-junit-migrator/pom.xml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        <version>0.8.1-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>testng-junit-migrator</artifactId>
+
+    <name>Picnic :: Error Prone Support :: TestNG JUnit Migrator</name>
+    <description>A TestNG to JUnit migration tool built on Error Prone Support</description>
+
+    <properties>
+        <maven.compiler.source>17</maven.compiler.source>
```
Comment (rickie) [inline testng-junit-migrator/pom.xml:19]: We can omit this.

File: testng-junit-migrator/pom.xml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            <artifactId>junit-jupiter-params</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (rickie) [inline testng-junit-migrator/pom.xml:96]: Make sure to call the `./format.sh` from a `picnic-shared-tools` repo.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGScanner.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  public ImmutableMap<ClassTree, TestNGMetadata> collectMetadaForEachClass(
+      CompilationUnitTree tree) {
+    scan(tree, null);
+    return metadataBuilder.build();
+  }
+
+  private static Optional<TestNGMetadata.AnnotationMetadata> getTestNGAnnotation(
+      Tree tree, VisitorState state) {
+    // XXX: I think there is a method to get a specific annotation from a tree. Not sure if that is
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGScanner.java:75]: Let's drop this one after our discussion :).

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigrationTest.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+            "import java.util.stream.Stream;",
+            "import org.junit.jupiter.api.DisplayName;",
+            "import org.junit.jupiter.api.MethodOrderer;",
+            "import org.junit.jupiter.api.Order;",
+            "import org.junit.jupiter.api.Test;",
+            "import org.junit.jupiter.api.TestMethodOrder;",
+            "import org.junit.jupiter.params.ParameterizedTest;",
+            "import org.junit.jupiter.params.provider.Arguments;",
+            "import org.junit.jupiter.params.provider.MethodSource;",
+            // XXX: Running `mvn clean install` (in this case still with `-Dverification.warn`) will
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigrationTest.java:145]: Please check out this comment 😉.

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/argument/PriorityArgumentMigratorTest.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package tech.picnic.errorprone.testngjunit.argument;
+
+import org.junit.jupiter.api.Test;
+
+final class PriorityArgumentMigratorTest {
+
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/argument/PriorityArgumentMigratorTest.java:6]:
```suggestion
```

Comment (rickie) [reply testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/argument/PriorityArgumentMigratorTest.java:6]: What does this one do 👀 ?

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.MethodTree;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+import tech.picnic.errorprone.bugpatterns.util.SourceCode;
+
+final class TestNGScannerTest {
+  // XXX: We are missing some tests here.
+  // - JUnit class that is already migrated.
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java:24]: Is this still up to date?

**Review by rickie — COMMENTED** [review]
SOme comments :D

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java (hunk 9 above)
Comment (oxkitsune) [reply testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java]: Now it is.

File: testng-junit-migrator/run-testng-junit-migration.sh [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+set -e -u -o pipefail
+
+# If this is not a Maven build, exit here to skip Maven-specific steps.
+if [ ! -f pom.xml ]; then
+  exit 0
+fi
+
+# Use the Maven Wrapper executable if present.
+MVN='mvn'
+if [ -x ./mvnw ]; then
```
Comment (rickie) [inline testng-junit-migrator/run-testng-junit-migration.sh:12]: We don't need to do this. We have this internally but there is a thread where someone mentioned it is not best practice.

File: testng-junit-migrator/run-testng-junit-migration.sh [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  -Dfrontend.skip \
+  -Dverification.skip
+
+echo "Counting JUnit tests run..."
+junit_results=$(${MVN} test | grep -n "Results:" -A 3 | grep -oP "Tests run: \K\d+(?=,)" | awk '{s+=$1} END {print s}')
+echo "Number of TestNG tests run: $testng_results"
+echo "Number of JUnit tests run: $junit_results"
+echo "Difference: $(expr ${testng_results} - ${junit_results})"
+
+echo "Finished migration steps!"
```
Comment (rickie) [inline testng-junit-migrator/run-testng-junit-migration.sh:116]: Will add a few tweaks to the script and strings :).

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/SourceCodeTest.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import javax.lang.model.element.Name;
+import org.junit.jupiter.api.Test;
+
+// Should be de-duplicated in the future, for example by depending on `error-prone-contrib`
+final class SourceCodeTest {
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/SourceCodeTest.java:21]: Hmm, we discussed this offline a few times. Let's drop it because it tests something we do not really use.

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGMatchersTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import com.google.errorprone.bugpatterns.BugChecker.AnnotationTreeMatcher;
+import com.google.errorprone.bugpatterns.BugChecker.MethodTreeMatcher;
+import com.google.errorprone.matchers.Description;
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.MethodTree;
+import org.junit.jupiter.api.Test;
+
+final class TestNGMatchersTest {
+  @Test
+  void nGAnnotation() {
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGMatchersTest.java:18]: There is some overlap here, we can combine the tests :). Will push a proposal.

**Review by rickie — COMMENTED** [review]
Flushing some comments :).

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    private static String createAnnotationDiagnosticMessage(
+        ClassTree classTree,
+        TestNGMetadata.AnnotationMetadata annotationMetadata,
+        VisitorState state) {
+      return String.format(
+          "class: %s arguments: { %s }",
+          classTree.getSimpleName(),
+          annotationMetadata.getArguments().entrySet().stream()
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java]: I'm wondering if there is a simpler way of just printing the data. Tried this:
```
      return String.format(
          classTree.getSimpleName(), annotationMetadata.getArguments().toString());
```
Results in:
 
> /A.java:18: error: [TestChecker] Class: A arguments: { {} }
  @Test
  ^
    (see https://errorprone.info/bugpattern/TestChecker)
/A.java:22: error: [TestChecker] Class: A arguments: { {description="foo"} }
  @Test(description = "foo")
  ^
    (see https://errorprone.info/bugpattern/TestChecker)
/A.java:26: error: [TestChecker] Class: A arguments: { {priority=1, description="foo"} }
  @Test(priority = 1, description = "foo")
  ^
    (see https://errorprone.info/bugpattern/TestChecker)
/A.java:30: error: [TestChecker] Class: A arguments: { {dataProvider="dataProviderTestCases"} }
  @Test(dataProvider = "dataProviderTestCases")
  ^
    (see https://errorprone.info/bugpattern/TestChecker)
expected to be true

Not that bad right? I propose to use that :).

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+                    (method, setupMethodType) ->
+                        state.reportMatch(
+                            buildDescription(method)
+                                .setMessage(
+                                    String.format(
+                                        "class: %s setupMethod: %s",
+                                        classTree.getSimpleName(), setupMethodType))
+                                .build()));
+
+            classTree.getMembers().stream()
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java:191]: Should we move this to a `matchMethodTree` method?

Comment (rickie) [reply testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java:191]: Have some thoughts about the setup of this class, maybe we can improve and make it cleaner :).

**Review by rickie — COMMENTED** [review]
Some thoughts :).

**Review by rickie — COMMENTED** [review]
Added two commits with quite some changes, PTAL. Didn't comment about all changes though.

**Comment by rickie** [comment]
Rebased, let's see what Sonar has to say 😋.

**Review by rickie — COMMENTED** [review]
So I added a few commits and rebased.
I mostly looked at polishing the code and making things a little simpler. It's another step in the direction of getting this merged :). Sorry I didn't add any commits, I just did a lot of refactoring while I was in the airplane twice 😄. The lack of internet made it hard to comment stuff haha.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestAnnotationAttribute.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package tech.picnic.errorprone.testngjunit;
+
+import static java.util.Arrays.stream;
+
+import java.util.Optional;
+
+/** The annotation argument kinds that are supported by the TestNG to JUnit Jupiter migration. */
+enum TestAnnotationAttribute {
+  DATAPROVIDER("dataProvider", new DataProviderAttributeMigrator()),
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestAnnotationAttribute.java:9]:
```suggestion
  DATA_PROVIDER("dataProvider", new DataProviderAttributeMigrator()),
```

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+ *   <li>{@code isEnabled}
+ *   <li>{@code expectedExceptions}
+ *   <li>{@code priority}
+ *   <li>{@code groups}
+ * </ul>
+ *
+ * This migration will also take care of any setup/teardown methods.
+ *
+ * <p>Note: As the {@code @BeforeAll} and {@code @AfterAll} methods in JUnit are required to be
+ * static, this <em>might</em> make breaking changes.
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGJUnitMigration.java:45]:
```suggestion
 * static, this <em>might</em> introduce breaking changes.
```

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/package-info.java [hunk 18]
```diff
@@ -0,0 +1,4 @@
+/** TestNG to JUnit migration using Error-Prone. */
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/package-info.java:1]: Reminder for myself to check this.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+import com.google.errorprone.annotations.Immutable;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodTree;
+import java.util.Optional;
+import tech.picnic.errorprone.util.SourceCode;
+
+/**
+ * A {@link Migrator} that migrates the {@code org.testng.annotations.Test#timeOut} argument. The
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java:13]:
```suggestion
 * A {@link Migrator} that migrates the {@code org.testng.annotations.Test#timeOut} attribute. The
```

**Review by rickie — COMMENTED** [review]
Some stuff for next time.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java (hunk 19 above)
Comment (rickie) [reply testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java:13]: (Also in other places I think)

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+ * attribute. The TestNG {@code org.testng.annotations.Test#timeOut} attribute is always in
+ * milliseconds, the JUnit variant {@code @Timeout} takes a value in seconds by default, hence we
+ * add the {@link java.util.concurrent.TimeUnit#MILLISECONDS} attribute.
+ */
+@Immutable
+final class TimeOutAttributeMigrator implements AttributeMigrator {
+  @Override
+  public Optional<SuggestedFix> migrate(
+      TestNGMetadata metadata,
+      TestNGMetadata.AnnotationMetadata annotation,
```
Comment (rickie) [inline testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java:21]: If we rename `AnnotationMetaData` to `TestNGAnnotationMetdata` we can omit the `TestNGMetadata` qualifier everywhere, I think that is cleaner, WDYT?

File: testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+            "",
+            "    // BUG: Diagnostic contains: Class: B attributes: {priority=1}",
+            "    @Test(priority = 1)",
+            "    public void nestedTestWithArguments() {}",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  // XXX: Here we need to add some edge cases for the DataProvider probably?
```
Comment (rickie) [inline testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java:78]: Here we wanted to document some of the edge cases right ?

**Review by rickie — COMMENTED** [review]
This is making a lot of progress :rocket:!

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGScanner.java [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+  @Override
+  public @Nullable Void visitMethod(MethodTree tree, TestNGMetadata.Builder builder) {
+    if (ASTHelpers.isGeneratedConstructor(tree)) {
+      return super.visitMethod(tree, builder);
+    }
+
+    DataProviderMigrator migrator = new DataProviderMigrator();
+    if (TESTNG_VALUE_FACTORY_METHOD.matches(tree, state) && migrator.canFix(tree)) {
+      builder
+          .dataProviderMetadataBuilder()
```
[the remark this answers is not in the document]
Comment (oxkitsune) [reply testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TestNGScanner.java:70]: This is the preferred way according to the AutoValue.Builder docs 😅.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/GroupsAttributeMigrator.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    SuggestedFix.Builder fix = SuggestedFix.builder().addImport("org.junit.jupiter.api.Tag");
+    groups.forEach(group -> fix.prefixWith(methodTree, String.format("@Tag(\"%s\")%n", group)));
+
+    return Optional.of(fix.build());
+  }
+
+  private static boolean isValidTagName(String tagName) {
+    return !tagName.isEmpty() && tagName.chars().noneMatch(Character::isISOControl);
```
[the remark this answers is not in the document]
Comment (oxkitsune) [reply testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/GroupsAttributeMigrator.java:43]: This is directly form the the [JUnit 5 docs](https://junit.org/junit5/docs/5.0.2/api/org/junit/jupiter/api/Tag.html): 
> Syntax Rules for Tags
> * A tag must not be blank.
> * A trimmed tag must not contain whitespace.
> * A trimmed tag must not contain ISO control characters.

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java (hunk 20 above)
Comment (oxkitsune) [reply testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/TimeOutAttributeMigrator.java:21]: So iiuc, we extract the `AnnotationMetadata` to a seperate class and call it `TestNGAnnotationMetadata`?

File: testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/DataProviderMigrator.java [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+
+  private DataProviderMigrator() {}
+
+  /**
+   * Tells whether the specified {@code DataProvider} can be migrated.
+   *
+   * @param methodTree The dataprovider methode tree.
+   * @return {@code true} if the data provider can be migrated or else {@code false}.
+   */
+  static boolean canFix(MethodTree methodTree) {
```
[the remark this answers is not in the document]
Comment (oxkitsune) [reply testng-junit-migrator/src/main/java/tech/picnic/errorprone/testngjunit/DataProviderMigrator.java:40]: Yes, but `DataProviderMigrator#canFix(MethodTree)` is used in a place where the arguments required for `#migrate` are not available.
